### ISSUE TITLE
Cannot connect to a SAP Cloud Platform OData service

### ISSUE THREAD

**Issue by s7oev** [SAP/python-pyodata#109]
Hello,

First of all, I want to mention that I am aware of the following code:
```
SERVICE_URL = 'https://odata.example.com/Secret.svc'

session = requests.Session()
session.auth = ('MyUser', 'MyPassword')

theservice = pyodata.Client(SERVICE_URL, session)
```

However, I have created an ABAP service following [this guide](https://developers.sap.com/group.abap-env-restful-managed.html) (more specifically, the service is created in step 3, Define and Expose a CDS-Based Travel Data Model). And, I got a service with the following structure:

<some_id>.abap-web.us10.hana.ondemand.com/sap/opu/odata/sap/<service_name>/

[Here's also the exact link](https://3995a331-5b33-4cab-8eeb-9ef5ad61985d.abap-web.us10.hana.ondemand.com/sap/opu/odata/sap/ZUI_SS_C_TRAVEL_M/) if that's necessary.

I cannot find any way to connect with it through PyOData. I imagine this is because SAP Cloud Platform uses OAuth2, while the example code above - HTTP Basic Auth. Still, I tried to input user and password data just to try, but that didn't work. I also managed to get an authorization token (from an xssec security context), which I think it might help, but can't see how to pass it properly.

Are there any suggestions for connecting to this service?

**Comment by filak-sap** [comment]
I didn't try it but you basically need to do OAuth2 authorization as described at:
https://requests-oauthlib.readthedocs.io/en/latest/oauth2_workflow.html#web-application-flow

and the use the variable oauth (in the 3rd step) instead for the variable session (from the code snippet).

Please, let me know if it works.

**Comment by s7oev** [comment]
Hey, Filak, thanks a lot for your response and suggestions.

Yes, I also tried using OAuth2Session already, tried again, unfortunately without luck. Best result I got was with this code:
```
from requests_oauthlib import OAuth2Session

client_id = r'{id here}'
client_secret = r'{secret here}'
redirect_uri = 'https://myapp778.cfapps.us10.hana.ondemand.com/sap/opu/odata/sap/ZUI_SS_C_TRAVEL_M/'

oauth = OAuth2Session(client_id, redirect_uri=redirect_uri,)
authorization_url, state = oauth.authorization_url('https://myapp778.cfapps.us10.hana.ondemand.com/')
print('Please go to %s and authorize access.' % authorization_url)

authorization_response = input('Enter the full callback URL')
print(authorization_response)
token = oauth.fetch_token(
        'https://myapp778.cfapps.us10.hana.ondemand.com/',
        authorization_response=authorization_response,
        client_secret=client_secret)
```
I do get a login link with a callback code, but after inserting it, I get:
```
raise MismatchingStateError()
oauthlib.oauth2.rfc6749.errors.MismatchingStateError: (mismatching_state) CSRF Warning! State not equal in request and respo
nse.
```

Either way, I've found a workaround by using an approuter to my service, where I can set the authentication to basic, and this is good enough for me. If it is of interest, this code in the xs-app.json does the trick:
 ```
   {
      "source": "^/sap/opu/odata/sap/ZUI_SS_C_TRAVEL_M/(.*)$",
      "target": "/sap/opu/odata/sap/ZUI_SS_C_TRAVEL_M/$1",
      "authenticationType": "basic",
      "service": "com.sap.cloud.abap",
      "endpoint": "abap",
      "csrfProtection": false
    }
```

Additionally, I've found that after I login, I can get the JSESSIONID from my cookie and pass it as follows:
```
    req_ses = requests.Session()
    req_ses.headers.update({'Cookie': jsessionid})
    theservice = pyodata.Client(SERVICE_URL, req_ses)
```
Which also works (even if authentication is OAuth2, not basic).

I would be interested if there is a way to connect with OAuth2, but it is proving too complex for me to find a way how. In case I do figure it out, I will share. 

So because of the possible workaround to use basic auth with an approuter, should I close the issue? Or are there plans to integrate PyOData with ABAP RESTful Programming Model, and it's better to leave it open?

(enjoy your weekend :) )

**Comment by filak-sap** [comment]
@mamiksik PTAL

**Comment by filak-sap** [comment]
@s7oev Lets leave this open. I have to make sure PyOData works with ABAP RESTful Programming Model.

**Comment by gsgxnet** [comment]
As this is still open after about half a year I assume it is not easy to get it going? 
Can we still hope for a solution. In my case it would be an S/4 onPrem.

**Comment by filak-sap** [comment]
We are currently short on time for adding enhancements. This issue shall be fixed once but nobody can tell when. We are always happy to merge a PR for it.

**Comment by s7oev** [comment]
Well, it took pretty much exactly one year, but I believe I finally have it solved :-) 

The code below authenticates for a RAP service coming from a SAP BTP (trial) instance. It's already a bit late, so for now I'll just copy-paste it as it is, and in the coming days I plan to test a few more times, and hopefully submit a PR with documentation.

The function below needs just:
1) the JSON service key for the ABAP environment on BTP (as a dictionary);
2) username & password for BTP login; and
3) the name of the service.

As a bonus, inside it also covers the CSRF token fetching, and returns a connected pyodata client with full CRUD ability.

The tricky part for connecting was that the RAP OData services from BTP do **not** accept the regular token provided from `/oauth/token` for whatever reason :/ but when you authenticate with both BTP username & password AND clientid & clientsecret (which are taken from the key), there are two tokens returned, one of which called `id_token`. This one does the trick.

```python
with open('key.txt', 'r') as f:
    KEY = f.read()
KEY = json.loads(KEY)

USER = "SAP BTP user"
PASSWORD = "SAP BTP password"
SERVICE_NAME = "what goes after (NOT INCLUDING) /sap/opu/odata/sap/; corresponds with service binding name"

def provision_client(key, user, password, service_name):
    token_url = key["uaa"]["url"]
    token_url += f"/oauth/token?grant_type=password&password={password}&username={user}"

    token_response = requests.post(token_url, auth=(key["uaa"]["clientid"], key["uaa"]["clientsecret"]))
    token_response = json.loads(token_response.text)

    token = token_response["id_token"]

    req_ses = requests.Session()
    req_ses.headers.update({'Authorization': f"Bearer {token}"})

    service_url = key["url"] + "/sap/opu/odata/sap/" + service_name
    response = req_ses.head(service_url, headers={'x-csrf-token': 'fetch'})
    token = response.headers.get('x-csrf-token', '')
    req_ses.headers.update({'x-csrf-token': token})

    client = pyodata.Client(service_url, req_ses)
    return client
```

**Comment by s7oev** [comment]
solved by f7274f8505cccc9bbfc1877d68d550c68949a6c0
